### PR TITLE
ZEC-28 Implement POST, PUT, and DELETE endpoints for products

### PR DESCRIPTION
**Pull request by AnnaMauhay**
ZEC-28 Implement POST, PUT, and DELETE endpoints for products

> https://mauhayannaliza.atlassian.net/browse/ZEC-28?atlOrigin=eyJpIjoiZTI0MzJkYWRhYTdjNGQ4MTlkYWRkNTZhMTUyNDE3MzMiLCJwIjoiaiJ9

1.a.  Create new product
> Screenshot 2024-01-15 at 12 36 23 PM

1.b. Fail to add existing product
> Screenshot 2024-01-15 at 12 38 26 PM

2.a. Edit Existing Product
> Screenshot 2024-01-15 at 12 39 36 PM

2.b. Fail to edit archived or product that is not created by the seller
> Screenshot 2024-01-15 at 12 41 24 PM

3.a. Archive existing/active product
> Screenshot 2024-01-15 at 12 43 47 PM

3.b. Fail to archive a non-active/ non existing product
> Screenshot 2024-01-15 at 12 44 57 PM

### CODE REVIEW

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java [hunk 1]
```diff
@@ -36,11 +45,63 @@ public ResponseEntity<Page<Product>> getProductsWithName(@RequestParam("keyword"
     }
     @GetMapping("/{id}")
     public ResponseEntity<?> getProductById(@PathVariable("id") int productId){
-        System.out.println(productId);
         try{
             return ResponseEntity.ok(productService.getProductById(productId));
         }catch (ProductNotFoundException e){
             return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
```
Comment (MenilV) [inline src/main/java/com/zalando/ecommerce/controller/ProductController.java:51]: It's good to have a JSON response here

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java [hunk 2]
```diff
@@ -36,11 +45,63 @@ public ResponseEntity<Page<Product>> getProductsWithName(@RequestParam("keyword"
...
+                                           @AuthenticationPrincipal UserDetails principal){
+        User user;
+        Product product;
+        try {
+            user = userService.getUserByEmail(principal.getUsername());
+            product = productService.addProduct(productRequest, user);
+        } catch (UserNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (DuplicateProductException e) {
+            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(e.getMessage());
```
Comment (MenilV) [inline src/main/java/com/zalando/ecommerce/controller/ProductController.java:66]: JSON

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java [hunk 3]
```diff
@@ -36,11 +45,63 @@ public ResponseEntity<Page<Product>> getProductsWithName(@RequestParam("keyword"
...
+        Product product;
+        try {
+            user = userService.getUserByEmail(principal.getUsername());
+            product = productService.updateProductDetails(productId, user, productRequest);
+        } catch (UserNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (ProductNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(e.getMessage());
+        }
+        return ResponseEntity.ok(product);
```
Comment (MenilV) [inline src/main/java/com/zalando/ecommerce/controller/ProductController.java:89]: JSON

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java [hunk 4]
```diff
@@ -36,11 +45,63 @@ public ResponseEntity<Page<Product>> getProductsWithName(@RequestParam("keyword"
...
+                                           @PathVariable("product-id") int productId){
+        User user;
+        Product product;
+        try {
+            user = userService.getUserByEmail(principal.getUsername());
+            product = productService.updateProductDetails(productId, user, productRequest);
+        } catch (UserNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (ProductNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(e.getMessage());
```
Comment (MenilV) [inline src/main/java/com/zalando/ecommerce/controller/ProductController.java:87]: JSON

**Review by MenilV — CHANGES_REQUESTED** [review]
A few changes to responses.

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java (hunk 4 above)
Comment (AnnaMauhay) [reply src/main/java/com/zalando/ecommerce/controller/ProductController.java]: Screenshot 2024-01-15 at 6 35 21 PM

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java (hunk 3 above)
Comment (AnnaMauhay) [reply src/main/java/com/zalando/ecommerce/controller/ProductController.java:90]: Screenshot 2024-01-15 at 6 34 32 PM

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java (hunk 2 above)
Comment (AnnaMauhay) [reply src/main/java/com/zalando/ecommerce/controller/ProductController.java]: Screenshot 2024-01-15 at 6 33 32 PM

File: src/main/java/com/zalando/ecommerce/controller/ProductController.java (hunk 1 above)
Comment (AnnaMauhay) [reply src/main/java/com/zalando/ecommerce/controller/ProductController.java]: Tested
Screenshot 2024-01-15 at 6 32 21 PM

**Review by faruka-startsteps — APPROVED** [review]

**Pull request merged**